### PR TITLE
Add `softmax_csr` implementation

### PR DESCRIPTION
**Pull request by DamianSzwichtenberg**
This PR adds forward and backward implementation of sparse softmax operation as defined [here](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/utils/softmax.py#L9).

In the `pytorch_geometric` implementation we cannot take advantage of model compilation when groups are defined via `ptr`. `softmax_csr` introduced here provides a well-performing kernel for such a scenario.

Performance boost (achieved on 28C, single socket machine):
~7x for forward pass
~8x for backward pass
Additionally, GAT training time was reduced by ~5%.

### CODE REVIEW

**Comment by DamianSzwichtenberg** [comment]
@pyg-team/intel-team Please take a look.

File: pyg_lib/ops/__init__.py [hunk 1]
```diff
@@ -332,6 +332,84 @@ def index_sort(
...
+
+        >>> src = torch.randn(4, 4)
+        >>> ptr = torch.tensor([0, 4])
+        >>> softmax(src, None, ptr)
+        tensor([[0.0157, 0.0984, 0.1250, 0.4523],
+                [0.1453, 0.2591, 0.5907, 0.2410],
+                [0.0598, 0.2923, 0.1206, 0.0921],
+                [0.7792, 0.3502, 0.1638, 0.2145]])
+    """
+    if src.dim() != 2 or not src.is_cpu or ptr is None or dim != 0:
```
Comment (kgajdamo) [inline pyg_lib/ops/__init__.py:403]: I wonder why ptr is optional because if you don't provide it, you get an error.

Comment (DamianSzwichtenberg) [reply pyg_lib/ops/__init__.py:403]: I wanted to make API in its final form, otherwise, each change here would require a change in `pytorch_geometric`. I'll add support for `index` in the near future.

Comment (DamianSzwichtenberg) [reply pyg_lib/ops/__init__.py]: @kgajdamo, after rethinking your suggestion, I decided to change the API and create a specialized `softmax_csr` operation that accepts `ptr` only. Rationale:
`torch.compile` gives nice results for `softmax` with groups defined via `index`, hence I don't see a reason to have a specialized kernel for that option.

**Comment by DamianSzwichtenberg** [comment]
@kgajdamo @rusty1s Please take a look. I made the softmax implementation a bit more general, so now it covers any `src` dimensionality as well as any given `dim`. I also restricted groups to be defined via `ptr` ([rationale here](https://github.com/pyg-team/pyg-lib/pull/264#discussion_r1381658840)).

**Comment by yanbing-j** [comment]
Hi @DamianSzwichtenberg , this PR looks good to me. The overall structure of softmax kernel with sparse input is similar with that in softmax kernel of dense input in PyTorch. With the sparsity, the performance boost is from parallelism, right?

And will this PR upstream to PyTorch later? Since there is no SparseCsr support for softmax yet in PyTorch.

File: pyg_lib/csrc/ops/cpu/softmax_kernel.cpp [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+at::Tensor softmax_csr_forward_kernel_impl(const at::Tensor& src,
+                                           const at::Tensor& groups,
+                                           const int64_t dim) {
+  auto out = at::zeros_like(src);
+
+  AT_DISPATCH_FLOATING_TYPES(
+      src.scalar_type(), "softmax_csr_forward_kernel_impl", [&] {
+        const auto n_groups = groups.size(0) - 1;
+        const auto n_heads = src.numel() / src.size(dim);
```
Comment (kgajdamo) [inline pyg_lib/csrc/ops/cpu/softmax_kernel.cpp:63]: I assume that n_heads should be of integer type, so src.numel() will always be divisible by src.size(dim)? Maybe I'd worth to add assert about that. Same for the backward.

Comment (DamianSzwichtenberg) [reply pyg_lib/csrc/ops/cpu/softmax_kernel.cpp:63]: I think it should be guaranteed. `numel()` can be represented as:
```python
numel_ = 1
for i in range(src.dim()):
    numel_ *= src.size(i)
```
As `numel` represents product of  `size`'s, it is divisible by them.

**Review by kgajdamo — APPROVED** [review]
Looks good to me.

File: pyg_lib/csrc/ops/cpu/softmax_kernel.cpp (hunk 2 above)
Comment (kgajdamo) [reply pyg_lib/csrc/ops/cpu/softmax_kernel.cpp:63]: Ok :)

File: pyg_lib/ops/__init__.py [hunk 3]
```diff
@@ -332,6 +331,62 @@ def index_sort(
     return torch.ops.pyg.index_sort(inputs, max_value)
 
 
+class Softmax(torch.autograd.Function):
+    @staticmethod
+    def forward(
```
Comment (rusty1s) [inline pyg_lib/ops/__init__.py:336]: Can we define the autograd function directly in C++?

Comment (DamianSzwichtenberg) [reply pyg_lib/ops/__init__.py:336]: Should be possible, will check.

Comment (DamianSzwichtenberg) [reply pyg_lib/ops/__init__.py:336]: #282
